packages/notebook/src/toc.ts

```diff
@@ -0,0 +1,704 @@
+
+      switch (model.type) {
+        case 'code': {
+          // Collapsing cells is incompatible with output headings
+          if (
+            !this.configuration.syncCollapseState &&
+            this.configuration.includeOutput
+          ) {
+            // Iterate over the code cell outputs to check for Markdown or HTML from which we can generate ToC headings...
+            const outputs = (model as CodeCellModel).outputs;
+            for (let j = 0; j < outputs.length; j++) {
+              const m = outputs.get(j);
+
+              let htmlType: string | null = null;
+              let mdType: string | null = null;
+
+              Object.keys(m.data).forEach(t => {
+                if (!mdType && ToCUtils.Markdown.isMarkdown(t)) {
+                  mdType = t;
+                } else if (!htmlType && ToCUtils.isHTML(t)) {
+                  htmlType = t;
+                }
+              });
+
+              // Parse HTML output only if trusted
+              if (htmlType) {
+                headings.push(
+                  ...ToCUtils.getHTMLHeadings(
+                    this.sanitizer.sanitize(m.data[htmlType] as string),
+                    this.configuration,
+                    documentLevels
+                  ).map(heading => {
+                    return {
+                      ...heading,
+                      cellRef: cell,
+                      index: [i, j],
+                      collapsed: false,
+                      isRunning: RunningStatus.Idle,
+                      type: HeadingType.HTML
+                    };
+                  })
+                );
+              } else if (mdType) {
+                headings.push(
+                  ...ToCUtils.Markdown.getHeadings(
+                    m.data[mdType] as string,
+                    this.configuration,
+                    documentLevels
+                  ).map(heading => {
+                    return {
+                      ...heading,
+                      cellRef: cell,
+                      index: [i, j],
+                      collapsed: false,
+                      isRunning: RunningStatus.Idle,
+                      type: HeadingType.Markdown
+                    };
+                  })
+                );
+              }
+            }
+          }
+
+          break;
+        }
+        case 'markdown': {
+          headings.push(
+            ...ToCUtils.Markdown.getHeadings(
+              cell.model.value.text,
+              this.configuration,
+              documentLevels
+            ).map((heading, index) => {
+              return {
+                ...heading,
+                cellRef: cell,
+                index: [i, 0],
+                collapsed:
+                  // If there are multiple headings, only collapse the first one
+                  this.configuration.syncCollapseState && index === 0
+                    ? (cell as MarkdownCell).headingCollapsed
+                    : false,
```

Comment:
Needs to change to be consistent with `Cell.headingInfo` and set the collapse state for the lowest-level heading within the cell.